### ISSUE TITLE
[java] OverrideBothEqualsAndHashCodeOnComparable should not be required for record classes

### ISSUE THREAD

**Issue by Juneezee** [pmd/pmd#6072]
**Affects PMD Version:** 

7.17.0

**Rule:**

OverrideBothEqualsAndHashCodeOnComparable: https://docs.pmd-code.org/pmd-doc-7.17.0/pmd_rules_java_errorprone.html#overridebothequalsandhashcodeoncomparable

**Description:**

Record classes have their `hashCode` and `equals` methods automatically implemented.

IntelliJ IDEA’s inspection ComparableImplementedButEqualsNotOverridden does not flag record classes as violations. Since OverrideBothEqualsAndHashCodeOnComparable was originally inspired by IDEA (https://github.com/pmd/pmd/issues/5837), PMD should also exclude record classes from this rule.

![Image](https://github.com/user-attachments/assets/86608a0a-c542-4bc3-8c45-0da0d7c6edac)

Reference: https://docs.oracle.com/javase/specs/jls/se17/html/jls-8.html#jls-8.10.3
Reference: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Record.html#equals(java.lang.Object)

**Code Sample demonstrating the issue:**

Adapted from https://www.jetbrains.com/help/inspectopedia/ComparableImplementedButEqualsNotOverridden.html:

```java
record Foo(int cm) implements Comparable<Foo> {
  @Override
  public int compareTo(Foo o) {
    if (cm == o.cm) {
      return 0;
    }
    return cm < o.cm ? -1 : 1;
  }
}

class Bar implements Comparable<Bar> {
  private final int cm;

  public Bar(int cm) {
    this.cm = cm;
  }

  @Override
  public int compareTo(Bar o) {
    if (cm == o.cm) {
      return 0;
    }
    return cm < o.cm ? -1 : 1;
  }
}
```

**Expected outcome:**

OverrideBothEqualsAndHashCodeOnComparable should not be required for record classes.

**Running PMD through:** *Gradle*

**Comment by UncleOwen** [comment]
I agree. `record`s already have a sane .equals()/.hashCode(). We shouldn't be telling users to add custom implementations if the default is fine.

**Comment by adangel** [comment]
The key point here is

> if the default is fine.

For the example given in this issue, it is fine: There is only one component, and this component is used within compareTo - so the implementation of compareTo "is consistent with equals".

But given the following example:

```java
import module java.base;

void main() {
  record MyRecord(String a, int b) implements Comparable<MyRecord> {
    public int compareTo(MyRecord other) {
      return b - other.b;
    }
  }

  List<MyRecord> data = Arrays.asList(new MyRecord("a", 1), new MyRecord("b", 1));
  IO.println("data size: " + data.size());

  Set<MyRecord> set = new HashSet<>(data);
  IO.println("set size: " + set.size());

  Set<MyRecord> sortedSet = new TreeSet<>(data);
  IO.println("sorted set: " + sortedSet.size());
}
```

If you run this (with java25), you'll get this output:

```
data size: 2
set size: 2
sorted set: 1
```

Which shows the effect of a compareTo implementation, that is _not_ consistent with equals. As described in [TreeSet](https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/util/TreeSet.html), the equality of two elements in a ordinary Set (e.g. HashSet) is based on equals, but the equality in a sorted Set (e.g. TreeSet) is based on compareTo. If these two are not consistent, you'll get different results depending on which Set implementation you use.

This is the whole point of the rule: Reminding developers to think about whether they use the elements in a sorted set and whether the compareTo and equals methods need to be consistent.

For the above example, an equals (and hashCode) method can be added to the record, which makes it consistent:

```java
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        return compareTo((MyRecord) o) == 0;
    }

    public int hashCode() {
      return b;
    }
```

Now, both Set implementations behave the same (which means, they behave consistent) - they correctly eliminate duplicates in the same way.

This is actually documented already in the rule (https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#overridebothequalsandhashcodeoncomparable):

> Note 2: This rule also reports records that implement Comparable. While for records, equals() and hashCode() are generated, you still must make sure that compareTo() is consistent with equals().


The question is now - what should a static code analyzer such as PMD do? Should we simply ignore records (records are nothing other than classes) (I guess, that is what JetBrains decided to do), should we improve the rule to really analyze the implementation of the compareTo method to figure out, if it is consistent with equals (which is a lot of effort and probably only feasible for simple compareTo methods), should we add a property so that users can deliberately ignore records, should we add an example (e.g. this example I've given) to make users understand the issue, should we add a separate rule "CompareToInRecordNotConsistentWithEquals?", should we suggest to suppress the rule in such cases ...?

@UncleOwen You suggested in #6081 to simple ignore records - which basically means, we let PMD err on the side of false-negatives rather than producing false-positives. I guess, if we do this, we need to create a new issue here that describes the false-negative, that we from now on won't be able to detect.

**Comment by UncleOwen** [comment]
> This is the whole point of the rule: Reminding developers to think about whether they use the elements in a sorted set and whether the compareTo and equals methods need to be consistent.

I believe that might be the core of the issue. Is "You should think about this" really enough? In the general case, I think the confidence rate (that is: number of true positives / number of violations) of this rule is very high. But in this one special case, records implementing Comparable, I don't think it is as high as we'd like it to be. Do we agree on that?

As to what to do next: Of course, fixing this rule properly or adding a new rule would be best. But I can't imagine how a proper solution would look like. Maybe verify that compareTo() looks at all the members? I don't know, looks brittle. But maybe worth investigating?

But until we have a proper solution, I prefer not triggering a violation over what we are doing currently. That's what #6081 is doing. If you'd like, I can make this configurable. But I would turn the violations off by default.

**Comment by adangel** [comment]
@jsotuyod @oowekyala @zbynek Should we err on the false positive side or the false negative side in this case? What do you think?

**Comment by zbynek** [comment]
I think comparable records should be only reported if they override exactly one of `equals` and `hashcode`.

**Comment by adangel** [comment]
> if they override exactly one of equals and hashcode

Then this should be reported by OverrideBothEqualsAndHashcode, shouldn't it? Ah, ok, no: that rule doesn't report on comparable types.

So, yes, that seems to be a reasonable middle ground.

Other suggestions:
* clearly say, when to use this rule: currently we way "if instances of these classes are used in collections". This should be "... **sorted** collections".
* remove this rule from the quickstart ruleset. As the rule makes only sense, if you use sorted collections... so - it doesn't apply in the general case.

@UncleOwen Could you update your PR #6081 to make sure, we have test cases for:
a) no violation:
```java
record Foo(int a) implements Comparable<Foo> {
  public int compareTo(Foo f) {
    return 0;
  }
}
```

b) one violation:
```java
record Foo(int a) implements Comparable<Foo> {
  public int compareTo(Foo f) {
    return 0;
  }
  public int hashCode() {
    return 0;
  }
}
```

c) one violation:
```java
record Foo(int a) implements Comparable<Foo> {
  public int compareTo(Foo f) {
    return 0;
  }
  public boolean equals(Object o) {
    return false;
  }
}
```

d) no violation:
```java
record Foo(int a) implements Comparable<Foo> {
  public int compareTo(Foo f) {
    return 0;
  }
  public int hashCode() {
    return 0;
  }
  public boolean equals(Object o) {
    return false;
  }
}
```

Thanks!
